curvefs/src/client/block_device_client.cpp

```diff
@@ -21,22 +21,199 @@
+                                          const std::string& owner,
+                                          BlockDeviceStat* statInfo) {
+    FileStatInfo fileStatInfo;
+    UserInfo userInfo(owner);
+    auto retCode = fileClient_->StatFile(filename, userInfo, &fileStatInfo);
+    if (retCode != LIBCURVE_ERROR::OK) {
+        LOG(ERROR) << "Stat file failed, retCode = " << retCode;
+        return CURVEFS_ERROR::FAILED;
+    }
+
+    statInfo->length = fileStatInfo.length;
+    statInfo->status = fileStatInfo.fileStatus;
+    return CURVEFS_ERROR::OK;
+}
+
+CURVEFS_ERROR BlockDeviceClientImpl::Read(char* buf,
+                                          off_t offset,
+                                          size_t length) {
+    if (!isOpen_ || fd_ < 0) {
+        return CURVEFS_ERROR::BAD_FD;
+    } else if (0 == length) {
+        return CURVEFS_ERROR::OK;
+    } else if (IsAligned(offset, length)) {
+        return AlignRead(buf, offset, length);
+    }
+
+    auto range = CalcAlignRange(offset, offset + length);  // [start, end)
+    off_t readStart = range.first;
+    off_t readEnd = range.second;
+    size_t readLength = readEnd - readStart;
+    std::unique_ptr<char[]> readBuffer(new (std::nothrow) char[readLength]);
+
+    auto retCode = AlignRead(readBuffer.get(), readStart, readLength);
+    if (retCode == CURVEFS_ERROR::OK) {
+        memcpy(buf, readBuffer.get() + (offset - readStart), length);
+    }
+
+    return retCode;
+}
+
+CURVEFS_ERROR BlockDeviceClientImpl::Write(const char* buf,
+                                           off_t offset,
+                                           size_t length) {
+    if (!isOpen_ || fd_ < 0) {
+        return CURVEFS_ERROR::BAD_FD;
+    } else if (0 == length) {
+        return CURVEFS_ERROR::OK;
+    } else if (IsAligned(offset, length)) {
+        return AlignWrite(buf, offset, length);
+    }
+
+    auto range = CalcAlignRange(offset, offset + length);  // [start, end)
+    off_t writeStart = range.first;
+    off_t writeEnd = range.second;
+    size_t writeLength = writeEnd - writeStart;
+    std::unique_ptr<char[]> writeBuffer(new (std::nothrow) char[writeLength]);
+
+    auto retCode = WritePadding(
+        writeBuffer.get(), writeStart, writeEnd, offset, length);
+    if (retCode == CURVEFS_ERROR::OK) {
+        memcpy(writeBuffer.get() + (offset - writeStart), buf, length);
+        retCode = AlignWrite(writeBuffer.get(), writeStart, writeLength);
+    }
+
+    return retCode;
+}
+
+CURVEFS_ERROR BlockDeviceClientImpl::WritePadding(char* writeBuffer,
+                                                  off_t writeStart,
+                                                  off_t writeEnd,
+                                                  off_t offset,
+                                                  size_t length) {
+    off_t readEnd = 0;
+    size_t writeLength = writeEnd - writeStart;
+
+    if (offset != writeStart) {  // Padding leading
```

Comment:
there is exists one case that can be optimized, e.g., IO_ALIGNED_BLOCK_SIZE is 4k, and the original write request's size is 2k, but it crosses a block boundary, like the last 1k of the first block and the first 1k of the second block.
so, you can just issue one read 8k request, rather than two 4k read requests.

and, read/write here are all synchronous operations, but I'm not sure that whether fuse support asynchronous processing, if it supports it's better to use the asynchronous version of read/write API.

BTW, it's not urgent to optimized this, so you can leave it.

Reply:
> there is exists one case that can be optimized, e.g., IO_ALIGNED_BLOCK_SIZE is 4k, and the original write request's size is 2k, but it crosses a block boundary, like the last 1k of the first block and the first 1k of the second block.
> so, you can just issue one read 8k request, rather than two 4k read requests.

This is indeed a good optimization that I have overlooked, i will improve it.

> and, read/write here are all synchronous operations, but I'm not sure that whether fuse support asynchronous processing, if it supports it's better to use the asynchronous version of read/write API.
> 
> BTW, it's not urgent to optimized this, so you can leave it.

I will to confirm whether the fuse support asynchronous processing and imporve it if it supports.